daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/DFDLXFunctions.scala

```diff
@@ -97,7 +97,59 @@ case class DFDLXRightShift(recipes: List[CompiledDPath], argType: NodeInfo.Kind)
     }
   }
 }
+case class DFDLXBitAnd(recipes: List[CompiledDPath], argType: NodeInfo.Kind) extends FNTwoArgs(recipes) {
+  override def computeValue(arg1: DataValuePrimitive, arg2: DataValuePrimitive, dstate: DState): DataValuePrimitive = {
+    argType match {
+      case NodeInfo.Long => arg1.getLong & arg2.getLong
+      case NodeInfo.Int => arg1.getInt & arg2.getInt
+      case NodeInfo.Short => (arg1.getShort & arg2.getShort).toShort
+      case NodeInfo.Byte => (arg1.getByte & arg2.getByte).toByte
+      case NodeInfo.UnsignedLong => arg1.getBigInt.and(arg2.getBigInt)
+      case NodeInfo.UnsignedInt => arg1.getLong & arg2.getLong
+      case NodeInfo.UnsignedShort => arg1.getInt & arg2.getInt
+      case NodeInfo.UnsignedByte => (arg1.getShort & arg2.getShort).toShort
+      //$COVERAGE-OFF$
+      case _ => Assert.invariantFailed(s"dfdlx:bitAND not supported")
+      // $COVERAGE-ON$
+    }
+  }
+}
 
+case class DFDLXBitOr(recipes: List[CompiledDPath], argType: NodeInfo.Kind) extends FNTwoArgs(recipes) {
+  override def computeValue(arg1: DataValuePrimitive, arg2: DataValuePrimitive, dstate: DState): DataValuePrimitive = {
+    argType match {
+      case NodeInfo.Long => arg1.getLong | arg2.getLong
+      case NodeInfo.Int => arg1.getInt | arg2.getInt
+      case NodeInfo.Short => (arg1.getShort | arg2.getShort).toShort
+      case NodeInfo.Byte => (arg1.getByte | arg2.getByte).toByte
+      case NodeInfo.UnsignedLong => arg1.getBigInt.or(arg2.getBigInt)
+      case NodeInfo.UnsignedInt => arg1.getLong | arg2.getLong
+      case NodeInfo.UnsignedShort => arg1.getInt | arg2.getInt
+      case NodeInfo.UnsignedByte => (arg1.getShort | arg2.getShort).toShort
+      //$COVERAGE-OFF$
+      case _ => Assert.invariantFailed(s"dfdlx:bitOR not supported")
+      // $COVERAGE-ON$
+    }
+  }
+}
+
+case class DFDLXBitNot(recipes: CompiledDPath, argType: NodeInfo.Kind) extends FNOneArg(recipes,argType) {
+  override def computeValue(arg1: DataValuePrimitive, dstate: DState): DataValuePrimitive = {
+    argType match {
+      case NodeInfo.Long => ~arg1.getLong
+      case NodeInfo.Int => ~arg1.getInt
+      case NodeInfo.Short => ~arg1.getShort
+      case NodeInfo.Byte => ~arg1.getByte
```

Comment:
Java's bitwise-not returns an Int, even if the type is short or a byte. But we need Short to stay Short, and Byte to stay Byte, so we need to case then. So these two lines should be:

```scala
case NodeInfo.Short => (~arg1.getShort).toShort
case NodeInfo.Byte => (~arg1.getByte).toByte
```
The unsigned cases looked right to me.